malariagen_data/ag3.py

```diff
@@ -8628,6 +8628,184 @@ def plot_h12_gwss(
+            sample set identifiers (e.g., ["AG1000G-BF-A", "AG1000G-BF-B"]) or a
+            release identifier (e.g., "3.0") or a list of release identifiers.
+        sample_query : str, optional
+            A pandas query string which will be evaluated against the sample
+            metadata e.g., "taxon == 'coluzzii' and country == 'Burkina Faso'".
+        color : str, optional
+            Identifies a column in the sample metadata with which the dendrogram leaves (haplotypes)
+            are coloured.
+        cohort_size : int, optional
+            If provided, randomly down-sample to the given cohort size.
+        random_seed : int, optional
+            Random seed used for down-sampling.
+        """
+        import plotly.express as px
+        from plotly.figure_factory import create_dendrogram
+        from scipy.cluster.hierarchy import linkage
+
+        debug = self._log.debug
+
+        debug("normalise parameters")
+        sample_sets = self._prep_sample_sets_arg(sample_sets=sample_sets)
+        region = self.resolve_region(region)
+        if isinstance(region, Region):
+            region = [region]
+
+        ds_haps = self.haplotypes(
+            region=region,
+            analysis=analysis,
+            sample_query=sample_query,
+            sample_sets=sample_sets,
+            cohort_size=cohort_size,
+            random_seed=random_seed,
+        )
+
+        gt = allel.GenotypeDaskArray(ds_haps["call_genotype"].data)
+        with self._dask_progress(desc="Load haplotypes"):
+            ht = gt.to_haplotypes().compute()
+
+        debug("load sample metadata")
+        df_samples = self.sample_metadata(
+            sample_sets=sample_sets, sample_query=sample_query
+        )
+        debug("align sample metadata with haplotypes")
+        phased_samples = ds_haps["sample_id"].values.tolist()
+        df_samples_phased = (
+            df_samples.set_index("sample_id").loc[phased_samples].reset_index()
+        )
+
+        debug("set up plotting options")
+        hover_data = [
+            "sample_id",
+            "partner_sample_id",
+            "sample_set",
+            "taxon",
+            "country",
+            "admin1_iso",
+            "admin1_name",
+            "admin2_name",
+            "location",
+            "year",
+            "month",
+        ]
+
+        plot_kwargs = dict(
+            template="simple_white",
+            hover_name="sample_id",
+            hover_data=hover_data,
+            render_mode="svg",
+        )
+
+        debug("special handling for taxon color")
+        if color == "taxon":
+            _setup_taxon_colors(plot_kwargs)
+
+        debug("apply any user overrides")
+        plot_kwargs.update(kwargs)
+
+        debug("Create dendrogram with plotly")
+        leaf_labels = np.arange(
+            ht.shape[1]
+        )  # set labels as the index which we extract to reorder metadata
```

Comment:
Nit: for longer comments like these, I'd usually put the comment on it's own line before the code it applies to. Otherwise it causes some slightly odd reformatting of code.